File: Safari-609.1.11.1/LayoutTests/http/tests/security/401-logout/401-logout.php
<?php
  if (!isset($_REQUEST['uid'])) {
    // Step 1 - navigate to a page that will make us remember credentials.
    echo "<script>\n";
    echo "if (!window.testRunner) {\n";
    echo "    document.write('This test only works as an automated one');\n";
    echo "    throw 0;\n";
    echo "}\n";
    echo "testRunner.waitUntilDone();\n";
    echo "testRunner.dumpAsText();\n";
    echo "testRunner.setHandlesAuthenticationChallenges(true)\n";
    echo "testRunner.setAuthenticationUsername('username')\n";
    echo "testRunner.setAuthenticationPassword('password')\n";
    echo "location = 'http://127.0.0.1:8000/security/401-logout/401-logout.php?uid=username';\n";
    echo "</script>\n";
  } else if (!isset($_SERVER['PHP_AUTH_USER']) || ($_REQUEST['uid'] != $_SERVER['PHP_AUTH_USER'])) {
    if (isset($_REQUEST['laststep'])) {
      // Step 4 - Credentials are no longer being sent
      echo "PASS";
      echo "<script>\n";
      echo "if (window.testRunner) {\n";
      echo "    testRunner.notifyDone();\n";
      echo "}\n";
      echo "</script>\n";
    } else {
      // Ask for credentials if there are none
      header('WWW-Authenticate: Basic realm="401-logout"');
      header('HTTP/1.0 401 Unauthorized');
    }
  } else {
    if (!isset($_REQUEST['logout'])) {
      // Step 2 - navigate to a page that will make us forget the credentials
      echo "<script>\n";
      echo "testRunner.setHandlesAuthenticationChallenges(false)\n";
      echo "location = 'http://127.0.0.1:8000/security/401-logout/401-logout.php?uid=username&logout=1';\n";
      echo "</script>\n";
    } else {
      // Step 3 - logout
      header('WWW-Authenticate: Basic realm="401-logout"');
      header('HTTP/1.0 401 Unauthorized');
      echo "<script>\n";
      echo "location = 'http://127.0.0.1:8000/security/401-logout/401-logout.php?uid=username&laststep=1';\n";
      echo "</script>\n";
    }
  }
?>

File: LayoutTests/http/tests/security/401-logout/401-logout.py
#!/usr/bin/env python3

import base64
import os
import sys

file = __file__.split(':/cygwin')[-1]
http_root = os.path.dirname(os.path.dirname(os.path.abspath(os.path.dirname(file))))
sys.path.insert(0, http_root)

from resources.portabilityLayer import get_request

username = base64.b64decode(os.environ.get('HTTP_AUTHORIZATION', ' Og==').split(' ')[1]).decode().split(':')[0]
request = get_request()

sys.stdout.write('Content-Type: text/html\r\n')

if request.get('uid', None) is None:
    # Step 1 - navigate to a page that will make us remember credentials.
    sys.stdout.write(
        '\r\n<script>\n'
        'if (!window.testRunner) {\n'
        '    document.write(\'This test only works as an automated one\');\n'
        '    throw 0;\n'
        '}\n'
        'testRunner.waitUntilDone();\n'
        'testRunner.dumpAsText();\n'
        'testRunner.setHandlesAuthenticationChallenges(true)\n'
        'testRunner.setAuthenticationUsername(\'username\')\n'
        'testRunner.setAuthenticationPassword(\'password\')\n'
        'location = \'http://127.0.0.1:8000/security/401-logout/401-logout.py?uid=username\';\n'
        '</script>\n'
    )

elif not username or request.get('uid', '') != username:
    if request.get('laststep', None) is not None:
        # Step 4 - Credentials are no longer being sent
        sys.stdout.write(
            '\r\nPASS<script>\n'
            'if (window.testRunner) {\n'
            '    testRunner.notifyDone();\n'
            '}\n'
            '</script>\n'
        )
    else:
        # Ask for credentials is there are none
        sys.stdout.write(
            'WWW-Authenticate: Basic realm="401-logout"\r\n'
            'status: 401\r\n\r\n'
        )
else:
    if request.get('logout', None) is None:
        # Step 2 - navigate to a page that will make us forget the credentials
        sys.stdout.write(
            '\r\n<script>\n'
            'testRunner.setHandlesAuthenticationChallenges(false)\n'
            'location = \'http://127.0.0.1:8000/security/401-logout/401-logout.py?uid=username&logout=1\';\n'
            '</script>\n'
        )
    else:
        # Step 3 - logout
        sys.stdout.write(
            'WWW-Authenticate: Basic realm="401-logout"\r\n'
            'status: 401\r\n\r\n'
            '<script>\n'
            'location = \'http://127.0.0.1:8000/security/401-logout/401-logout.py?uid=username&laststep=1\';\n'
            '</script>\n'
        )
